### PR TITLE
PackageManagerDownloadWorker refactor: step 1

### PR DESCRIPTION
**Pull request by tiegz**
Cleanup, refactor, and wrangle an N+1.

### CODE REVIEW

File: app/models/package_manager/base.rb [hunk 1]
```diff
@@ -88,35 +88,8 @@ def self.platform_name(platform)
     end
 
     def self.map_project(project)
-      mapped_project = mapping(project)
-      mapped_project = mapped_project.delete_if { |_key, value| value.blank? } if mapped_project.present?
-      mapped_project
-    end
-
-    def self.save(project, sync_versions: true)
```
Comment (tiegz) [inline app/models/package_manager/base.rb:96]: There's not really a reason for this to be outside of the `update()` method, so I consolidated them.

File: app/models/package_manager/base.rb [hunk 2]
```diff
@@ -217,26 +205,25 @@ def self.new_names
...
-                 dependencies(name, version.number, mapped_project)
+                 dependencies(name, db_version.number, mapped_project)
                rescue StandardError
                  []
                end
-        next unless deps&.any? && version.dependencies.empty?
 
         deps.each do |dep|
-          next if dep[:project_name].blank? || version.dependencies.find_by_project_name(dep[:project_name])
+          next if dep[:project_name].blank? || db_version.dependencies.any? { |d| d.project_name == dep[:project_name] }
```
Comment (tiegz) [inline app/models/package_manager/base.rb:219]: Here's the db win in this PR: we've avoided an N+1 above with `db_project.versions.includes(:dependencies)`, but then reintroduced the N+1 and made it N^2+1 ( by doing `version.dependencies.find_by_project_name` for each dep for each version 😆 . This should fix.

File: app/models/package_manager/base.rb [hunk 3]
```diff
@@ -132,42 +105,57 @@ def self.update_version(name, version)
...
     end
 
-    def self.finalize_dbproject(dbproject)
-      dbproject.reload
-      dbproject.download_registry_users
-      dbproject.last_synced_at = Time.now
-      dbproject.save
-      dbproject
+    def self.finalize_db_project(db_project)
+      db_project.reload
```
Comment (jsonperl) [inline app/models/package_manager/base.rb:134]: Do we need the reload?

File: app/models/package_manager/base.rb [hunk 4]
```diff
@@ -132,42 +105,57 @@ def self.update_version(name, version)
...
+      db_project.reformat_repository_url unless db_project.new_record?
+      db_project.update(mapped_project.except(:name, :releases, :versions, :version, :dependencies, :properties))
+
+      if self::HAS_VERSIONS && sync_versions
+        versions(project, db_project.name)
+          .each { |v| add_version(db_project, v) }
+          .tap { |vs| deprecate_versions(db_project, vs) }
+      end
+
+      save_dependencies(mapped_project) if self::HAS_DEPENDENCIES
```
Comment (jsonperl) [inline app/models/package_manager/base.rb:157]: Looks like we do this in line 111 too?

File: app/models/package_manager/base.rb [hunk 5]
```diff
@@ -132,42 +105,57 @@ def self.update_version(name, version)
...
       proj = project(name)
-      save(proj, sync_versions: sync_versions) if proj.present?
+      return unless project.present?
+
+      mapped_project = map_project(project)
+      return false unless mapped_project.present?
+
+      db_project = Project.find_or_initialize_by({ name: mapped_project[:name], platform: db_platform })
+      db_project.reformat_repository_url unless db_project.new_record?
+      db_project.update(mapped_project.except(:name, :releases, :versions, :version, :dependencies, :properties))
```
Comment (tiegz) [inline app/models/package_manager/base.rb:149]: Simplified this stanza that was originally in `self.save()`

File: app/models/package_manager/base.rb (hunk 4 above)
Comment (jsonperl) [reply app/models/package_manager/base.rb:157]: Ah that's the version specific side of this.

File: app/models/package_manager/base.rb [hunk 6]
```diff
@@ -132,42 +105,57 @@ def self.update_version(name, version)
...
+
+      if self::HAS_VERSIONS && sync_versions
```
Comment (jsonperl) [inline app/models/package_manager/base.rb:151]: Wonder if a later cleanup would be relying on all the stuff in update_version and just iterating through versions in the update method.

File: app/models/package_manager/base.rb (hunk 3 above)
Comment (tiegz) [reply app/models/package_manager/base.rb:134]: I bet we don't -- in fact I could probably just move this logic into the main `update()` method and try to consolidate all the UPDATEs.

File: app/models/package_manager/base.rb (hunk 2 above)
Comment (jsonperl) [reply app/models/package_manager/base.rb:219]: Now *that* is what I'm talking about! 💪

File: app/models/package_manager/base.rb (hunk 3 above)
Comment (jsonperl) [reply app/models/package_manager/base.rb:134]: Do it!

Comment (tiegz) [reply app/models/package_manager/base.rb:134]: (I'm gonna push this to a followup PR to keep these PRs chunked)

Comment (jsonperl) [reply app/models/package_manager/base.rb:134]: Screen Shot 2021-02-17 at 2 18 07 PM

**Review by jsonperl — APPROVED** [review]

**Pull request merged**